src/mapbuffer.cpp

```diff
@@ -130,20 +130,22 @@ void mapbuffer::save( bool delete_after_save )
         // A segment is a chunk of 32x32 submap quads.
         // We're breaking them into subdirectories so there aren't too many files per directory.
         // Might want to make a set for this one too so it's only checked once per save().
-        std::stringstream segment_path;
+        std::stringstream dirname;
         tripoint segment_addr = overmapbuffer::omt_to_seg_copy( om_addr );
-        segment_path << map_directory.str() << "/" << segment_addr.x << "." <<
+        dirname << map_directory.str() << "/" << segment_addr.x << "." <<
                      segment_addr.y << "." << segment_addr.z;
-        assure_dir_exist( segment_path.str().c_str() );
 
         std::stringstream quad_path;
-        quad_path << segment_path.str() << "/" << om_addr.x << "." <<
+        quad_path << dirname.str() << "/" << om_addr.x << "." <<
                   om_addr.y << "." << om_addr.z << ".map";
 
         // delete_on_save deletes everything, otherwise delete submaps
         // outside the current map.
-        save_quad( quad_path.str(), om_addr, submaps_to_delete,
-                   delete_after_save || om_addr.z != g->levz ||
+#ifndef ZLEVELS
+        delete_after_save |= om_addr.z != g->levz;
```

Comment:
You probably want a local (to the loop) variable here. Otherwise the first submap that has a z-level != g->levz will set it to `true` and all further iterations of that loop will be affected. Once `delete_after_save` is set to true, it will never be set back to false.

Theoretical  example:
Start with `delete_after_save = false` (the default).
Start in the shelter, go down and up again. The submaps are ordered by their coordinates, the first iterated submap has a `z-level != g->levz` and that would make `delete_after_save` true before `save_quad` is called even once.
Therefor `save_quad` will always be called with `delete_after_save = true` and *all* submaps will be deleted. Even those that are still in use.